integrations/graal/mp-native-image-extension/src/main/java/io/helidon/integrations/graal/mp/nativeimage/extension/WeldFeature.java

```diff
@@ -0,0 +1,308 @@
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.enterprise.context.spi.CreationalContext;
+import javax.enterprise.inject.Any;
+import javax.enterprise.inject.Default;
+import javax.enterprise.inject.Instance;
+import javax.enterprise.inject.spi.Bean;
+import javax.enterprise.inject.spi.CDI;
+import javax.enterprise.inject.spi.InjectionPoint;
+import javax.json.Json;
+import javax.json.JsonArray;
+import javax.json.JsonObject;
+import javax.json.JsonReaderFactory;
+
+import io.helidon.integrations.graal.nativeimage.extension.NativeConfig;
+
+import com.oracle.svm.core.annotate.AutomaticFeature;
+import org.graalvm.nativeimage.hosted.Feature;
+import org.graalvm.nativeimage.hosted.RuntimeReflection;
+import org.jboss.weld.bean.proxy.ClientProxyFactory;
+import org.jboss.weld.bean.proxy.ClientProxyProvider;
+import org.jboss.weld.manager.BeanManagerImpl;
+import org.jboss.weld.util.Proxies;
+
+/**
+ * An automatic feature for native-image to
+ *  register Weld specific stuff.
+ */
+@AutomaticFeature
+public class WeldFeature implements Feature {
+    private static final boolean ENABLED = NativeConfig.option("weld.enable-feature", true);
+    private static final boolean TRACE = NativeConfig.option("weld.trace", false);
+    private static final boolean WARN = NativeConfig.option("weld.warn", false);
+
+    @Override
+    public boolean isInConfiguration(IsInConfigurationAccess access) {
+        return ENABLED;
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public void duringSetup(DuringSetupAccess access) {
+        Class<?> beanManagerClass = access.findClassByName("org.jboss.weld.manager.BeanManagerImpl");
+        Set<Class<?>> processed = new HashSet<>();
+        Set<Set<Type>> processedExplicitProxy = new HashSet<>();
+        Set<Object> processedBeanManagers = Collections.newSetFromMap(new IdentityHashMap<>());
+        List<WeldProxyConfig> weldProxyConfigs = weldProxyConfigurations(access);
+
+        trace(() -> "Weld feature");
+
+        access.registerObjectReplacer((obj) -> {
+            if (beanManagerClass.isInstance(obj) && processedBeanManagers.add(obj)) {
+                try {
+                    BeanManagerImpl bm = (BeanManagerImpl) obj;
+                    ClientProxyProvider cpp = bm.getClientProxyProvider();
+
+                    String contextId = bm.getContextId();
+                    List<Bean<?>> beans = bm.getBeans();
+                    // TODO it should be sufficient to create proxy classes, no need to actually select stuff
+                    // the selected stuff is cached and may cause trouble in runtime (configuration!!!)
+                    iterateBeans(bm, cpp, processed, beans);
+
+                    weldProxyConfigs.forEach(proxy -> {
+                        initializeProxy(access,
+                                        processedExplicitProxy,
+                                        contextId,
+                                        proxy.beanClass,
```

Comment:
Classes related to native-image require refactoring. I will create a follow up issue to resolve this. 
I you have a better idea how to initialize the proxy classes, please let me know - even if it would require modification of Weld code.

Reply:
Follow up issue: https://github.com/oracle/helidon/issues/1273